### PR TITLE
[DO NOT MERGE] D0ix via platform device runtime pm

### PR DESCRIPTION
**Pull request by keyonjie**
This series create a platform device which is used for D0ix runtime power management, it should be merged after #1232 is merged.

@ranj063 @plbossart I changed to put the platform device to Intel HDA specific(similar to DMIC platform device), and doing put/get in hda pcm open/close.

### CODE REVIEW

File: sound/soc/sof/intel/pm_d0ix.c [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+#define sof_d0ix_resume NULL
+#endif
+
+static int pm_d0ix_probe(struct platform_device *pdev)
+{
+	/* allow runtime_pm */
+	pm_runtime_set_autosuspend_delay(&pdev->dev, SND_SOF_D0I3_DELAY_MS);
+	pm_runtime_use_autosuspend(&pdev->dev);
+
+	pm_runtime_allow(&pdev->dev);
```
Comment (ranj063) [inline sound/soc/sof/intel/pm_d0ix.c:50]: @keyonjie this is not needed for platform devices only for PCI devices. pm_runtime_enable() will do the job

File: sound/soc/sof/intel/hda-pcm.c [hunk 2]
```diff
@@ -235,5 +249,15 @@ int hda_dsp_pcm_close(struct snd_sof_dev *sdev,
...
 	/* unbinding pcm substream to hda stream */
 	substream->runtime->private_data = NULL;
+
+	spcm = snd_sof_find_spcm_dai(sdev, rtd);
+	if (!spcm)
+		return -EINVAL;
+
+	if (!spcm->stream[substream->stream].d0i3_compatible) {
+		pm_runtime_mark_last_busy(&hdev->d0ix_dev->dev);
+		pm_runtime_put_autosuspend(&hdev->d0ix_dev->dev);
```
Comment (ranj063) [inline sound/soc/sof/intel/hda-pcm.c:259]: @keyonjie your assumption is that when the last non-d0i3-compatible stream is closed, the DSP should enter D0i3?
But this should be the case only if a d0i3-compatible stream is active isnt it?

Comment (plbossart) [inline sound/soc/sof/intel/hda-pcm.c:259]: I've stated this before, but we should not mix layers. Each device should take care of its own power management, you are abusing the pm_runtime framework here.
I am not ok with this.

**Comment by ranj063** [comment]
@plbossart @keyonjie At the outset, the idea of using a separate platform device simplifies the handling of when to enter the D0i3. But I agree that we'd be messing the parent-child PM relationship by explicitly callingget/put from the parent PCI dev. 

On the other hand, the solution provided by Keyon earlier to use a delayed work to check for active D0i3 compatible streams is a bit straightforward. But it also comes with the perils of having to manage this work.

Alternatively, I had a different thought about using the platform driver. We could have 2 separate component drivers for the D0i3 compatible and non-D0i3 compatible dai drivers. The D0i3 compatible  component device would be tied to the top-level PCI device whereas the non-D0i3 compatible component device would be tied with the newly created platform device. This way the runtime PM get/put sync calls would be managed automatically by the core when the PCM's are opened or closed. When none of the non-D0i3 compatible streams are active, the platform device would get runtime suspended and its runtime suspend callback can make the decision to enter D0i3 if there are any D0i3 compatible streams active. When no streams are active, the PCI device would suspend normally. Just thinking out loud here. Comments welcome

**Comment by plbossart** [comment]
> @plbossart @keyonjie At the outset, the idea of using a separate platform device simplifies the handling of when to enter the D0i3. But I agree that we'd be messing the parent-child PM relationship by explicitly callingget/put from the parent PCI dev.
> 
> On the other hand, the solution provided by Keyon earlier to use a delayed work to check for active D0i3 compatible streams is a bit straightforward. But it also comes with the perils of having to manage this work.

The main problem I have with the current solution is that the get/put cases are only handled on open/close. 
The alternate solution implemented in the SKL driver was to track IPC calls, and setup D0ix when a timer expired after the last IPC and some streams were opened but could tolerate D0ix. A new IPC would cancel on-going timer and restart one when the IPC was complete. That sounds to me like a better model of the interaction with the firmware.
Only considering the open-close cases as transitions doesn't look very good to me.

> Alternatively, I had a different thought about using the platform driver. We could have 2 separate component drivers for the D0i3 compatible and non-D0i3 compatible dai drivers. The D0i3 compatible component device would be tied to the top-level PCI device whereas the non-D0i3 compatible component device would be tied with the newly created platform device. This way the runtime PM get/put sync calls would be managed automatically by the core when the PCM's are opened or closed. When none of the non-D0i3 compatible streams are active, the platform device would get runtime suspended and its runtime suspend callback can make the decision to enter D0i3 if there are any D0i3 compatible streams active. When no streams are active, the PCI device would suspend normally. Just thinking out loud here. Comments welcome

Not able to follow, you just left me in the dirt.

**Comment by plbossart** [comment]
To go further on my previous note, if there is an application that opens the device, but sends a trigger say 10s later (after the pm_runtime timeout), then by that time the device may transitioned to D0i3 but the IPC is sent anyway. It looks broken to me. It's really each IPC that should wake up the device if needed, and the transition to D0ix happens at a later time.

**Comment by ranj063** [comment]
> To go further on my previous note, if there is an application that opens the device, but sends a trigger say 10s later (after the pm_runtime timeout), then by that time the device may transitioned to D0i3 but the IPC is sent anyway. It looks broken to me. It's really each IPC that should wake up the device if needed, and the transition to D0ix happens at a later time.

@plbossart I think this is very good point. Looking at the code today, something like the IPC flood test would fail when the DSP in D0i3 because we'd never wake it up to D0 to send the IPC.

**Comment by plbossart** [comment]
@keyonjie I believe this PR is in abandoned state? can you close it if we have no use for this?

File: sound/soc/sof/pm.c [hunk 3]
```diff
@@ -422,3 +438,41 @@ int snd_sof_suspend(struct device *dev)
...
+
+	dev_dbg(sdev->dev, "setting ADSP D0 substate:%d\n", d0_substate);
+
+	/* Todo: destroy/reload unused pipelines/widgets in FW */
+
+	/* do platform specific set_state */
+	ret = snd_sof_dsp_set_power_state(sdev, d0_substate);
+	if (ret < 0) {
+		dev_err(sdev->dev,
+			"error: d0i3_exit ipc error %d\n", ret);
```
Comment (lyakh) [inline sound/soc/sof/pm.c:456]: is this message correct - is this really necessarily a "d0i3_exit error?"

File: sound/soc/sof/pm.c [hunk 4]
```diff
@@ -422,3 +438,41 @@ int snd_sof_suspend(struct device *dev)
...
+	if (d0_substate == SOF_DSP_D0I3)
+		flags = SOF_PM_PG_STREAMING;
+	else
+		flags = SOF_PM_PPG | SOF_PM_PCG;
+
+	/* sending pm_gate IPC */
+	ret = sof_send_pm_gate_ipc(sdev, flags);
+	if (ret < 0) {
+		dev_err(sdev->dev,
+			"error: d0i3_exit ipc error %d\n", ret);
```
Comment (lyakh) [inline sound/soc/sof/pm.c:469]: ditto, and would be good to have different error lines to distinguish what has failed

File: sound/soc/sof/intel/cnl.c [hunk 5]
```diff
@@ -150,14 +150,36 @@ static void cnl_ipc_dsp_done(struct snd_sof_dev *sdev)
 				CNL_DSP_REG_HIPCCTL_DONE);
 }
 
+static bool cnl_is_ipc_compact(struct snd_sof_ipc_msg *msg)
+{
+	/* the compact IPC subset might be extended later */
+	if (msg->header == (SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_GATE))
+		return true;
+	else
+		return false;
```
Comment (lyakh) [inline sound/soc/sof/intel/cnl.c:159]: please make this
```
return msg->header == (SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_GATE);
```

**Comment by plbossart** [comment]
> Because of @plbossart <https://github.com/plbossart> 's recent comment 
> to this PR it jumped up in my notification list and without looking at 
> what actually brought it up I started reviewing it :-/ Will be a lesson 
> for me, but anyway not to throw my comments away...

Hehe, I will add you to all my comments now :-)

**Comment by keyonjie** [comment]
let me close this one and will not re-open it until D0i3 on S0 is required.

**Comment by plbossart** [comment]
> let me close this one and will not re-open it until D0i3 on S0 is required.

well it is required if you want to test on all our CI platforms w/o S0ix...

**Pull request closed without merging**